sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ChannelEnumerableSubscription.cs

```diff
@@ -0,0 +1,255 @@
+        /// <param name="unsubscribeCallbackAsync">The callback function to invoke to unsubscribe.</param>
+        /// <param name="cancellationToken">The <see cref="System.Threading.CancellationToken"/> instance to signal the request to cancel reading from the channel for enumeration.</param>
+        ///
+        public ChannelEnumerableSubscription(ChannelReader<T> reader,
+                                             TimeSpan? maximumWaitTime,
+                                             Func<Task> unsubscribeCallbackAsync,
+                                             CancellationToken cancellationToken)
+        {
+            Guard.ArgumentNotNull(nameof(reader), reader);
+            Guard.ArgumentNotNull(nameof(unsubscribeCallbackAsync), unsubscribeCallbackAsync);
+
+            if (maximumWaitTime.HasValue)
+            {
+                Guard.ArgumentNotNegative(nameof(maximumWaitTime), maximumWaitTime.Value);
+            }
+
+            ChannelReader = reader;
+            MaximumWaitTime = maximumWaitTime;
+            UnsubscribeCallbackAsync = unsubscribeCallbackAsync;
+            CancellationToken = cancellationToken;
+        }
+
+        /// <summary>
+        ///   Builds an asynchronous enumerator based on the event channel.
+        /// </summary>
+        ///
+        /// <param name="cancellationToken">The cancellation token.</param>
+        ///
+        /// <returns>The asynchronous enumerator to use for iterating over events.</returns>
+        ///
+        public IAsyncEnumerator<T> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(ChannelEnumerableSubscription<T>));
+            }
+
+            return new ChannelEnumerator(EnumerateChannel(ChannelReader, MaximumWaitTime, Random.Value, CancellationToken)
+                .GetAsyncEnumerator(cancellationToken), DisposeAsync);
+        }
+
+        /// <summary>
+        ///   Performs the tasks needed to clean up the enumerable, including
+        ///   attempting to unsubscribe from the associated subscription.
+        /// </summary>
+        ///
+        ///
+        public async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            await UnsubscribeCallbackAsync().ConfigureAwait(false);
+            _disposed = true;
+        }
+
+        /// <summary>
+        ///   Enumerates the events as they become available in the associated channel.
+        /// </summary>
+        ///
+        /// <param name="reader">The reader for the channel from which events are to be sourced.</param>
+        /// <param name="maximumWaitTime">The maximum amount of time to wait to for an event to be available before emitting an empty item; if <c>null</c>, empty items will not be emitted.</param>
```

Comment:
> The maximum amount of time to wait to for an event to be available before emitting an empty item; if <c>null</c>, empty items will not be emitted.

It's an interesting pattern we don't usually use. What is the use case?

Reply:
The key concept is that an event stream is not a finite source; it is an infinite source that you have no deterministic ending for.   A couple of additional points of reasoning: 

- It's a common pattern in messaging systems to be able to emit "punctuation" at a measured period as a health check to determine if no events are being sent or if your infrastructure is in a bad state.

- It's a pattern established for Event Hubs with receiving giving the option in track one whether by batch or by handler.

- It allows control to return to the loop for the option to cease without the need for something in the background to trigger the cancellation token or upfront knowledge to set a specific time for cancellation.   Commonly, you see a pattern of "if we get XX amount of empty events in a row, then we want to suspend processing or add a delay before we start receiving again."

Reply:
Why can't we stop the enumeration in this case?

Reply:
We could, but it's not the goal.   If that's what we were after, then setting a timeout on the cancellation token would serve the purpose.  The goal is to allow consumers to say "I want to receive something every X period, even if its empty.   Give me events as you get them, but if we have no activity in that time period, emit an empty."